flash/image/classification/integrations/baal.py

```diff
@@ -0,0 +1,143 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from copy import deepcopy
+from functools import partial
+from typing import Any, Callable, Dict, Optional
+
+import numpy as np
+import torch
+from pytorch_lightning.loops import Loop
+from pytorch_lightning.loops.fit_loop import FitLoop
+from pytorch_lightning.trainer.progress import Progress
+from pytorch_lightning.trainer.states import RunningStage, TrainerFn, TrainerStatus
+from torch.nn.modules.batchnorm import _BatchNorm
+
+import flash
+from flash.core.utilities.imports import _BAAL_AVAILABLE, requires
+
+if _BAAL_AVAILABLE:
+    from baal.active import ActiveLearningDataset
+    from baal.active.heuristics import AbstractHeuristic, BALD
+
+
+class ActiveLearningTrainer(flash.Trainer):
+    def __init__(self, *args, **kwags):
+        super().__init__(*args, **kwags)
+
+        active_learning_loop = ActiveLearningLoop(label_epoch_frequency=1, heuristics=BALD(reduction=np.mean))
+        active_learning_loop.connect(self.fit_loop)
+        self.fit_loop = active_learning_loop
+        active_learning_loop.trainer = self
+
+
+class ActiveLearningLoop(Loop):
+    @requires("baal")
+    def __init__(self, label_epoch_frequency: int, heuristics: "AbstractHeuristic", inference_iteration: int = 2):
+        super().__init__()
+        self.label_epoch_frequency = label_epoch_frequency
+        self.fit_loop: Optional[FitLoop] = None
+        self.state_dict: Optional[Dict[str, torch.Tensor]] = None
+        self.progress = Progress()
+        self.heuristics = heuristics
+        self._should_continue: bool = False
+        self.inference_iteration = inference_iteration
+
+    def __getattr__(self, key):
+        if key not in self.__dict__:
+            return getattr(self.fit_loop, key)
+        return self.__dict__[key]
+
+    def connect(self, fit_loop: FitLoop):
+        self.fit_loop = fit_loop
+        self.max_epochs = self.fit_loop.max_epochs
+        self.fit_loop.max_epochs = self.label_epoch_frequency
+
+    def _to_active_dataset(self, dataset):
+        return ActiveLearningDataset(dataset, labelled=torch.arange(len(dataset) - 10))
```

Comment:
is there a reason that we force what percentage of the data is labelled?

Reply:
This is a hack while the PR is in progress. 
The length is being checked and it would break if no labelled data are provided. 
This would definitely be changed to the right approach following your example / codebase.

I was actually thinking to do the following:

```py
while not done:
    reset model
    train on train_dataset
    compute scores on predict train_dataset
    request for labelisation
    merge train_dataset + predict labelled data
    predict_dataset : Subset of non label data
    store current labellisation state in the loop.

move labelled data to train folder
```

Reply:
If you believe another approach should be taken, please share the pseudo code and I will implement it :) 

Enabling non labelled dataset for training might requires more work on the data loading part, but definitely feasible. 

Option 1:

Provide a way for the datamodule to load unlabelled data directly

```py
ImageClassificationData.from_folders(
    train_folder=...
    trainer_non_labelled_folder=...
)
```

Option 2:

Add support for unlabelled folder with contains non-labelled images.

I would be curious to hear the best practices on data management for active learning.

Reply:
@tchaton I think that on the datamodule side things should remain the same, from an API perspective. Internally, the class should split the training set into a pool and a labelled set. That's usually what one expects.

Maybe the process can look like
```
while not done
    if labelled data:  # <- in the first iteration, if you have data train otherwise use the initialized model
        train on labelled
    predict the pool dataset
    compute scores
    etc   
```
Usually (even if tiny and chosen at random) a labelled budget is the starting point of active learning. However, having the possibility of allowing the model to choose the instances to label from the very beginning can be interesting (think of pretrained models that you want to adapt to downstream task: they might know what they want)

Reply:
I agree with @pietrolesci  on this. I think, prior to training it is good to give the option to the user to input which samples are labelled. Otherwise, they would use the `label_randomly` functionality of `ActiveDataset` to label a few samples and then start the training. It totally makes sense for the training to not support an empty training set (no labels in the dataset). `ActiveDataset` is automatically taking care of what is used for the training and what is used for the `pool` so that part doesn't need handling unless you would need to update another dataset with the newly created `ActiveDataset.pool` at each step.